jax/_src/pjit.py

```diff
@@ -1341,10 +1400,43 @@ def pjit_staging_rule(trace, *args, **params):
       all(_is_unspecified(o) for o in params["out_shardings"])):
     jaxpr = params['jaxpr']
     return core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
-  else:
+  elif not config.jax_dynamic_shapes:
     return trace.default_process_primitive(pjit_p, args, params)
+  else:
+    source_info = source_info_util.current()
```

Comment:
nit: add `assert config.jax_dynamic_shapes` here?

Reply:
Hrm, it seems a bit redundant given the `elif` branch above is `elif not config.jax_dynamic_shapes`. That is, the structure is
```python
if ...:
  ...
elif not X:
  ...
else:
  # you want to put an `assert X` here?
  ...
```

If we didn't have the first `if` branch it'd basically be:

```python
if not X:
 ...
else:
  # put an `assert X` here?
  ...
```

Is the assertion really needed?

Reply:
Not needed.. just for readability purpose. Up to you! (I don't have strong feelings)

(I missed the `elif not config.jax_dynamic_shapes`  thing on the first read)

Reply:
Maybe I'll just switch the latter two branches. I was "putting the complicated thing at the end" but maybe it's less clear, as you say.

Reply:
Done! It's more readable now, on the important axis of being able to easily evaluate "this is only hit with dynamic shapes". Good idea.